### ISSUE TITLE
[Feature Request] Option for Self Contained Publish of Class Library to Contain Runtime

### ISSUE THREAD

**Issue by ben3441** [dotnet/sdk#12303]
My scenario is I am deploying a C++ wrapper to a .net core library, the .net core library will be loaded into the native process using [.net core hosting](https://docs.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting).

In this case I don't have an executable to publish although my library is dependent on the .net core runtime in order for the .net core hosting to work. To work around this I have added a decoy console application in my solution which I publish in order to get the runtime into my output directory then I delete the decoy app.

My request is for there to be an optional argument on the dotnet publish command to force the runtime to be included even if the published project is not an executable.

**Comment by ThaDaVos** [comment]
Same request here - I am working on extension dlls for Clarion based apps - sadly it's x86 only, so can't use the Native OAT Single File thing - as there will be multiple DLLS created I rather ship the runtime with them to ease inclusion in projects.

I also noticed, if I set `<SelfContained>false</SelfContained>` to true - my DLL won't even work and just error out (I am using DNNE to create an DLL with exports) - this all because the `runtimeconfig.json` is different because of `<SelfContained>false</SelfContained>`

**Comment by marcpopMSFT** [comment]
@agocke thoughts on this particular scenario? I'm not sure a self-contained library makes sense for most customers but there is a niche situation here.

**Comment by agocke** [comment]
As the owner of DNNE, I think @AaronRobinsonMSFT might have a better sense here. The Native AOT scenario seems to work fine, so it's just the conventional hosting side.

**Comment by AaronRobinsonMSFT** [comment]
> I also noticed, if I set <SelfContained>false</SelfContained> to true - my DLL won't even work and just error out (I am using DNNE to create an DLL with exports)

I'm not sure why setting `<SelfContained>false</SelfContained>` changes anything, does not setting it have a different behavior? @ThaDaVos I wouldn't set that for DNNE. There is a "self-contained" scenario for DNNE, but until the SDK provides the support being described here, it requires manually copying over the required binaries.

**Comment by agocke** [comment]
The problem with this scenario from the runtime perspective is that having two CLRs in the same app isn't supported.

**Comment by AaronRobinsonMSFT** [comment]
> The problem with this scenario from the runtime perspective is that having two CLRs in the same app isn't supported.

Agreed. There is definitely a conceptual issue with the scenario, but the underlying request is sound if the user developer is constructing a plug-in system for .NET components. Visual Studio is an example where the host is unmanaged and wants to own a plug-in system around .NET. Of course this can be abused, but asking users to manually collect the needed self-contained assets is annoying and I assume should be relatively simple to permit. Add a big ugly warning that can't be suppressed so they know the problems they are likely to face.

**Comment by agocke** [comment]
In the user scenario above, would it make more sense to publish the runtime with the plug-in host, rather than a plugin (which it sounds like this feature request describes)?

**Comment by AaronRobinsonMSFT** [comment]
> publish the runtime with the plug-in host

The plug-in host is a library, not a plug-in itself. Recall that the application is unmanaged and it uses a .NET library to author a .NET plug-in system. That means the .NET library that is going to be the host needs to be self-contained. It is precisely what DNNE is doing now.
